app/src/wavevr/cpp/DeviceDelegateWaveVR.cpp

```diff
@@ -191,6 +202,44 @@ struct DeviceDelegateWaveVR::State {
 
   }
 
+  void UpdateFoveatedLevel() {
+    if (!WVR_IsRenderFoveationSupport()) {
+      VRB_LOG("This Wave device doesn't support Foveation Render.");
+      return;
+    }
+
+    if (!leftFBOIndex || !rightFBOIndex) {
+      return;
+    }
+
+    uint32_t quality = (renderMode == device::RenderMode::StandAlone)
+                        ? standaloneFoveatedLevel : immersiveFoveatedLevel;
+
+    if (quality == currentFoveatedLevel) {
+      return;
+    }
+
+    if (quality == 0) {
+      // This is not working, we have to restart the app to reset.
+      WVR_RenderFoveation(false);
+    } else {
+      WVR_RenderFoveation(true);
+      // Mapping foveated level (1~3) to WVR_PeripheralQuality (high~low).
+      WVR_PeripheralQuality peripheralQuality =
+              static_cast<WVR_PeripheralQuality>(WVR_PeripheralQuality_High-(quality - 1));
+
+      WVR_RenderFoveationParams_t foveated;
+      foveated.focalX = foveated.focalY = 0.0f;
+      foveated.fovealFov = foveatedFov;
+      foveated.periQuality = peripheralQuality;
+      WVR_TextureParams_t eyeTexture = WVR_GetTexture(leftTextureQueue, leftFBOIndex);
+      WVR_PreRenderEye(WVR_Eye_Left, &eyeTexture, &foveated);
+      eyeTexture = WVR_GetTexture(rightTextureQueue, rightFBOIndex);
+      WVR_PreRenderEye(WVR_Eye_Right, &eyeTexture, &foveated);
```

Comment:
This doesn't look right. From the documentation, `WVR_PreRenderEye()` needs to be called each time before we render to the texture. This looks like it just gets the current texture and does it once when the level changes. This should probably go in either `DeviceDelegateWaveVR::StartFrame()` or in `DeviceDelegateWaveVR::BindEye()`.

Reply:
Removed the if-checking of level changes. As you said, we should just set the foveated level when rendering scene. I have already put it at `DeviceDelegateWaveVR::StartFrame()` -> `DeviceDelegateWaveVR::UpdateFoveatedLevel()` at my previous patch, so it should be good.